### PR TITLE
Small refactoring around drain - move drain logic to separate function.

### PR DESCRIPTION
**Pull request by mwielgus**
cc: @piosz @fgrzadkowski

### CODE REVIEW

File: pkg/kubectl/cmd/drain.go [hunk 1]
```diff
@@ -195,17 +195,55 @@ func (o *DrainOptions) RunDrain() error {
...
+	if len(daemonSetPodNames) > 0 {
+		fmt.Fprintf(o.out, "WARNING: Skipping %s\n", unmanagedMsg([]string{}, daemonSetPodNames, false))
+	}
+
+	return pods, nil
+}
+
+// GetPodsForDeletionOnNodeDrain returns pods that should be deleted on node drain as well as some extra information
+// about possibly problematic pods (unreplicated and deamon sets).
+func GetPodsForDeletionOnNodeDrain(client *client.Client, nodename string, decoder runtime.Decoder, removeUnderplicated bool,
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: Suggest making this private

File: pkg/kubectl/cmd/drain.go [hunk 2]
```diff
@@ -195,17 +195,55 @@ func (o *DrainOptions) RunDrain() error {
...
+	ignoreDeamonSet bool) (pods []api.Pod, unreplicatedPodNames []string, daemonSetPodNames []string, finalError error) {
+
+	pods = []api.Pod{}
+	unreplicatedPodNames = []string{}
+	daemonSetPodNames = []string{}
+	podList, err := client.Pods(api.NamespaceAll).List(api.ListOptions{FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": nodename})})
+	if err != nil {
+		return []api.Pod{}, []string{}, []string{}, err
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: `return pods, unreplicatedPodNames, daemonSetPodNames, err`

File: pkg/kubectl/cmd/drain.go [hunk 3]
```diff
@@ -195,17 +195,55 @@ func (o *DrainOptions) RunDrain() error {
...
+}
+
+// GetPodsForDeletionOnNodeDrain returns pods that should be deleted on node drain as well as some extra information
+// about possibly problematic pods (unreplicated and deamon sets).
+func GetPodsForDeletionOnNodeDrain(client *client.Client, nodename string, decoder runtime.Decoder, removeUnderplicated bool,
+	ignoreDeamonSet bool) (pods []api.Pod, unreplicatedPodNames []string, daemonSetPodNames []string, finalError error) {
+
+	pods = []api.Pod{}
+	unreplicatedPodNames = []string{}
+	daemonSetPodNames = []string{}
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: aren't these named return args already?

Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: never mind

File: pkg/kubectl/cmd/drain.go [hunk 4]
```diff
@@ -220,19 +258,19 @@ func (o *DrainOptions) GetPodsForDeletion() ([]api.Pod, error) {
 		if found {
 			// Now verify that the specified creator actually exists.
 			var sr api.SerializedReference
-			if err := runtime.DecodeInto(o.factory.Decoder(true), []byte(creatorRef), &sr); err != nil {
-				return pods, err
+			if err := runtime.DecodeInto(decoder, []byte(creatorRef), &sr); err != nil {
+				return []api.Pod{}, []string{}, []string{}, err
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: `return pods, unreplicatedPodNames, daemonSetPodNames, err`

File: pkg/kubectl/cmd/drain.go (hunk 2 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: actually, we can change `err` to `finalError` and just `return`

File: pkg/kubectl/cmd/drain.go (hunk 4 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: ditto, `return`

File: pkg/kubectl/cmd/drain.go (hunk 1 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: `removeUnderplicated`: did you mean `removeUnreplicated` (typo)?

Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: why not use `force`?

File: pkg/kubectl/cmd/drain.go [hunk 5]
```diff
@@ -270,34 +308,14 @@ func (o *DrainOptions) GetPodsForDeletion() ([]api.Pod, error) {
...
-
-	if len(unreplicatedPodNames) > 0 {
-		fmt.Fprintf(o.out, "WARNING: About to delete these %s\n", unmanagedMsg(unreplicatedPodNames, []string{}, false))
-	}
-	if len(daemonSetPodNames) > 0 {
-		fmt.Fprintf(o.out, "WARNING: Skipping %s\n", unmanagedMsg([]string{}, daemonSetPodNames, false))
-	}
-
-	return pods, nil
+	return pods, unreplicatedPodNames, daemonSetPodNames, nil
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: ditto; `return`

File: pkg/kubectl/cmd/drain.go [hunk 6]
```diff
@@ -195,17 +195,55 @@ func (o *DrainOptions) RunDrain() error {
...
+	pods, unreplicatedPodNames, daemonSetPodNames, err := GetPodsForDeletionOnNodeDrain(
+		o.client,
+		o.nodeInfo.Name,
+		o.factory.Decoder(true),
+		o.Force,
+		o.IgnoreDaemonsets,
+	)
 	if err != nil {
-		return pods, err
+		return []api.Pod{}, err
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: `return pods, err`

File: pkg/kubectl/cmd/drain.go [hunk 7]
```diff
@@ -195,17 +195,55 @@ func (o *DrainOptions) RunDrain() error {
...
+		return []api.Pod{}, errors.New(unmanagedMsg([]string{}, daemonSetPodNames, true))
+	case unreplicatedErrors && !daemonSetErrors:
+		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, []string{}, true))
+	}
+
+	if len(unreplicatedPodNames) > 0 {
+		fmt.Fprintf(o.out, "WARNING: About to delete these %s\n", unmanagedMsg(unreplicatedPodNames, []string{}, false))
+	}
+	if len(daemonSetPodNames) > 0 {
+		fmt.Fprintf(o.out, "WARNING: Skipping %s\n", unmanagedMsg([]string{}, daemonSetPodNames, false))
```
Comment (janetkuo) [inline pkg/kubectl/cmd/drain.go]: Can we combine these two?

File: pkg/kubectl/cmd/drain.go (hunk 1 above)
Comment (mwielgus) [inline pkg/kubectl/cmd/drain.go]: The whole purpose of this refactoring is to make it public :).

Other comments applied.

File: pkg/kubectl/cmd/drain.go (hunk 2 above)
Comment (mwielgus) [inline pkg/kubectl/cmd/drain.go]: Done

File: pkg/kubectl/cmd/drain.go (hunk 4 above)
Comment (mwielgus) [inline pkg/kubectl/cmd/drain.go]: I don't want to return non-empty arrays in case of an error.

File: pkg/kubectl/cmd/drain.go (hunk 5 above)
Comment (mwielgus) [inline pkg/kubectl/cmd/drain.go]: I don't want to return non-empty arrays in case of an error.

File: pkg/kubectl/cmd/drain.go (hunk 7 above)
Comment (mwielgus) [inline pkg/kubectl/cmd/drain.go]: This code was already there.

File: pkg/kubectl/cmd/drain.go (hunk 2 above)
Comment (mwielgus) [inline pkg/kubectl/cmd/drain.go]: However, for clarity I would be better to return empty arrays explicitley.

**Comment by janetkuo** [comment]
LGTM now. Thanks!

**Comment by bgrant0607** [comment]
cc @mml

**Comment by piosz** [comment]
Bumped priority, cluster autoscaler in a blocker for 1.3

**Pull request merged**